### PR TITLE
Levelup City Implementation 

### LINKED ISSUE #2267: [Bug] i can'nt offer the Anemoculus to the statue
help me, please i can'nt offer the Anemoculus to the statue
I play normally and haven't used any commands yet 
Screenshot_34
Screenshot_33
Screenshot_36

### PR DESCRIPTION
**Pull request by Phong940253**
## Description

See #2267.
Levelup City Implementation.
requires placing the [StatuePromoteExcel.json](https://gitlab.com/phong940253/GC-Resources/-/blob/e4a52a3bcf1c08d1b8b54dcfcf82937fc9ed9123/Resources/ExcelBinOutput/StatuePromoteExcelConfigData.json) file in /resources/ExcelBinOutput

## Issues fixed by this PR
Fix level Statue of the Seven always shows 1.

## Type of changes

- [x] Bug fix
- [x] New feature 
- [ ] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Comment by akbaryahya** [comment]
hmm have you tried this on the old account? (after Implementation) because it doesn't seem to be able to handle the null `getProperty` n `setProperty`, maybe it should have been if their old account should have been set with a defultf value.

get data
```
00:59:30 <ERROR:GameServerPacketHandler> Error handle
java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "java.util.Map.get(Object)" is null
        at emu.grasscutter.game.player.Player.getProperty(Player.java:815)
        at emu.grasscutter.game.managers.SotSManager.getCurrentLevel(SotSManager.java:257)
        at emu.grasscutter.server.packet.send.PacketGetSceneAreaRsp.<init>(PacketGetSceneAreaRsp.java:27)
        at emu.grasscutter.server.packet.recv.HandlerGetSceneAreaReq.handle(HandlerGetSceneAreaReq.java:17)
        at emu.grasscutter.server.game.GameServerPacketHandler.handle(GameServerPacketHandler.java:87)
        at emu.grasscutter.server.game.GameSession.handleReceive(GameSession.java:190)
        at emu.grasscutter.server.game.GameSessionManager$1.lambda$handleReceive$0(GameSessionManager.java:79)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
        at io.netty.channel.DefaultEventLoop.run(DefaultEventLoop.java:54)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
```		
set data
```
03:26:46 <ERROR:GameServerPacketHandler> Error handle
java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "java.util.Map.get(Object)" is null
        at emu.grasscutter.game.player.Player.setPropertyWithSanityCheck(Player.java:1651)
        at emu.grasscutter.game.player.Player.setProperty(Player.java:807)
        at emu.grasscutter.game.managers.SotSManager.setCurrentCrystal(SotSManager.java:240)
        at emu.grasscutter.game.managers.SotSManager.levelUpSotS(SotSManager.java:312)
        at emu.grasscutter.server.packet.recv.HandlerLevelupCityReq.handle(HandlerLevelupCityReq.java:19)
        at emu.grasscutter.server.game.GameServerPacketHandler.handle(GameServerPacketHandler.java:87)
        at emu.grasscutter.server.game.GameSession.handleReceive(GameSession.java:190)
        at emu.grasscutter.server.game.GameSessionManager$1.lambda$handleReceive$0(GameSessionManager.java:79)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
        at io.netty.channel.DefaultEventLoop.run(DefaultEventLoop.java:54)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by NotThorny** [comment]
Shouldn't this be using [CityLevelupConfigData.json](https://gitlab.com/YuukiPS/GC-Resources/-/blob/3.5/Resources/ExcelBinOutput/CityLevelupConfigData.json)? Looking at your new excel it appears you've at least looked over the original, I believe it would make more sense to use existing resources instead of creating something new that won't exist in future dumps, and also keeps it backwards compatible with all earlier versions and resources.

One other thing is that while this is adding the stamina, it's not adding the correct value of stamina. The player property MAX_STAMINA starts at 10000 (100 stamina) with max 24000 (240 stamina). You're adding the in-game stamina value of 8, which gives the user no actual new stamina. You need to add 800 to MAX_STAMINA property (8 stamina) to match values with the in-game stamina shown.

Then last is old accounts will null pointer with the manager, probably need a check for that.
Other than those appears to be working properly.

**Comment by Phong940253** [comment]
> hmm have you tried this on the old account? (after Implementation) because it doesn't seem to be able to handle the null `getProperty` n `setProperty`, maybe it should have been if their old account should have been set with a defultf value.
> 
> get data
> 
> ```
> 00:59:30 <ERROR:GameServerPacketHandler> Error handle
> java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "java.util.Map.get(Object)" is null
>         at emu.grasscutter.game.player.Player.getProperty(Player.java:815)
>         at emu.grasscutter.game.managers.SotSManager.getCurrentLevel(SotSManager.java:257)
>         at emu.grasscutter.server.packet.send.PacketGetSceneAreaRsp.<init>(PacketGetSceneAreaRsp.java:27)
>         at emu.grasscutter.server.packet.recv.HandlerGetSceneAreaReq.handle(HandlerGetSceneAreaReq.java:17)
>         at emu.grasscutter.server.game.GameServerPacketHandler.handle(GameServerPacketHandler.java:87)
>         at emu.grasscutter.server.game.GameSession.handleReceive(GameSession.java:190)
>         at emu.grasscutter.server.game.GameSessionManager$1.lambda$handleReceive$0(GameSessionManager.java:79)
>         at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
>         at io.netty.channel.DefaultEventLoop.run(DefaultEventLoop.java:54)
>         at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
>         at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
>         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
>         at java.base/java.lang.Thread.run(Thread.java:833)
> ```
> 
> set data
> 
> ```
> 03:26:46 <ERROR:GameServerPacketHandler> Error handle
> java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "java.util.Map.get(Object)" is null
>         at emu.grasscutter.game.player.Player.setPropertyWithSanityCheck(Player.java:1651)
>         at emu.grasscutter.game.player.Player.setProperty(Player.java:807)
>         at emu.grasscutter.game.managers.SotSManager.setCurrentCrystal(SotSManager.java:240)
>         at emu.grasscutter.game.managers.SotSManager.levelUpSotS(SotSManager.java:312)
>         at emu.grasscutter.server.packet.recv.HandlerLevelupCityReq.handle(HandlerLevelupCityReq.java:19)
>         at emu.grasscutter.server.game.GameServerPacketHandler.handle(GameServerPacketHandler.java:87)
>         at emu.grasscutter.server.game.GameSession.handleReceive(GameSession.java:190)
>         at emu.grasscutter.server.game.GameSessionManager$1.lambda$handleReceive$0(GameSessionManager.java:79)
>         at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
>         at io.netty.channel.DefaultEventLoop.run(DefaultEventLoop.java:54)
>         at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
>         at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
>         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
>         at java.base/java.lang.Thread.run(Thread.java:833)
> ```

Oh, this happened on the old account so I'll fix it soon

**Comment by akbaryahya** [comment]
> > hmm have you tried this on the old account? (after Implementation) because it doesn't seem to be able to handle the null `getProperty` n `setProperty`, maybe it should have been if their old account should have been set with a defultf value.
> > get data
> > ```
> > 00:59:30 <ERROR:GameServerPacketHandler> Error handle
> > java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "java.util.Map.get(Object)" is null
> >         at emu.grasscutter.game.player.Player.getProperty(Player.java:815)
> >         at emu.grasscutter.game.managers.SotSManager.getCurrentLevel(SotSManager.java:257)
> >         at emu.grasscutter.server.packet.send.PacketGetSceneAreaRsp.<init>(PacketGetSceneAreaRsp.java:27)
> >         at emu.grasscutter.server.packet.recv.HandlerGetSceneAreaReq.handle(HandlerGetSceneAreaReq.java:17)
> >         at emu.grasscutter.server.game.GameServerPacketHandler.handle(GameServerPacketHandler.java:87)
> >         at emu.grasscutter.server.game.GameSession.handleReceive(GameSession.java:190)
> >         at emu.grasscutter.server.game.GameSessionManager$1.lambda$handleReceive$0(GameSessionManager.java:79)
> >         at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
> >         at io.netty.channel.DefaultEventLoop.run(DefaultEventLoop.java:54)
> >         at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
> >         at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
> >         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> >         at java.base/java.lang.Thread.run(Thread.java:833)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > set data
> > ```
> > 03:26:46 <ERROR:GameServerPacketHandler> Error handle
> > java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "java.util.Map.get(Object)" is null
> >         at emu.grasscutter.game.player.Player.setPropertyWithSanityCheck(Player.java:1651)
> >         at emu.grasscutter.game.player.Player.setProperty(Player.java:807)
> >         at emu.grasscutter.game.managers.SotSManager.setCurrentCrystal(SotSManager.java:240)
> >         at emu.grasscutter.game.managers.SotSManager.levelUpSotS(SotSManager.java:312)
> >         at emu.grasscutter.server.packet.recv.HandlerLevelupCityReq.handle(HandlerLevelupCityReq.java:19)
> >         at emu.grasscutter.server.game.GameServerPacketHandler.handle(GameServerPacketHandler.java:87)
> >         at emu.grasscutter.server.game.GameSession.handleReceive(GameSession.java:190)
> >         at emu.grasscutter.server.game.GameSessionManager$1.lambda$handleReceive$0(GameSessionManager.java:79)
> >         at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
> >         at io.netty.channel.DefaultEventLoop.run(DefaultEventLoop.java:54)
> >         at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
> >         at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
> >         at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> >         at java.base/java.lang.Thread.run(Thread.java:833)
> > ```
> 
> Oh, this happened on the old account so I'll fix it soon

btw in `levelUpSotS` maybe you should also add `player.getQuestManager().queueEvent(QuestContent.QUEST_CONTENT_CITY_LEVEL_UP, cityId, areaId);` (idk this is true or areaId should be level) to trigger quest finish (46901-46902) (Offer the Anemoculus to the Statue /ll)

**Comment by Phong940253** [comment]
> Shouldn't this be using [CityLevelupConfigData.json](https://gitlab.com/YuukiPS/GC-Resources/-/blob/3.5/Resources/ExcelBinOutput/CityLevelupConfigData.json)? Looking at your new excel it appears you've at least looked over the original, I believe it would make more sense to use existing resources instead of creating something new that won't exist in future dumps, and also keeps it backwards compatible with all earlier versions and resources.
> 
> One other thing is that while this is adding the stamina, it's not adding the correct value of stamina. The player property MAX_STAMINA starts at 10000 (100 stamina) with max 24000 (240 stamina). You're adding the in-game stamina value of 8, which gives the user no actual new stamina. You need to add 800 to MAX_STAMINA property (8 stamina) to match values with the in-game stamina shown.
> 
> Then last is old accounts will null pointer with the manager, probably need a check for that. Other than those appears to be working properly.



> Shouldn't this be using [CityLevelupConfigData.json](https://gitlab.com/YuukiPS/GC-Resources/-/blob/3.5/Resources/ExcelBinOutput/CityLevelupConfigData.json)? Looking at your new excel it appears you've at least looked over the original, I believe it would make more sense to use existing resources instead of creating something new that won't exist in future dumps, and also keeps it backwards compatible with all earlier versions and resources.
> 
> One other thing is that while this is adding the stamina, it's not adding the correct value of stamina. The player property MAX_STAMINA starts at 10000 (100 stamina) with max 24000 (240 stamina). You're adding the in-game stamina value of 8, which gives the user no actual new stamina. You need to add 800 to MAX_STAMINA property (8 stamina) to match values with the in-game stamina shown.
> 
> Then last is old accounts will null pointer with the manager, probably need a check for that. Other than those appears to be working properly.

I've looked in resources but can't find data related to the costItems of each city upgrade level. So I simulated [WeaponPromoteExcelConfigData.json](https://gitlab.com/YuukiPS/GC-Resources/-/blob/fbb7ab0c23f0f62494fb7aadd669cc11cf71114e/Resources/ExcelBinOutput/WeaponPromoteExcelConfigData.json) . With costItems based on the original game and reward information taken from [RewardExcelConfigData.json](https://gitlab.com/YuukiPS/GC-Resources/-/blob/ab0a5fd3df3d618dcae62092a89285cf3566020b/Resources/ExcelBinOutput/RewardExcelConfigData.json). If you have any other way to save, please suggest me.

About MAX_STAMINA, I thought the stamina base was 100, and 10000 was due to the unlimited stamina hack so I didn't pay attention to this problem. haha. I will fix it soon.

**Comment by NotThorny** [comment]
In CityLevelupConfigData.json, consumeItem is the item taken ("costItem" equivalent), and it includes amount. Example here for level 2 (as first level is given by default) where the item being consumed has id included and 1 is needed.
```
"level": 2,
"consumeItem": {
            "itemId": 107001,
            "itemNum": 1
        },
```

File: src/main/java/emu/grasscutter/game/managers/SotSManager.java [hunk 1]
```diff
@@ -208,4 +214,140 @@ public void run() {
...
+                return player.getProperty(PlayerProperty.PROP_PLAYER_ANEMOCULUS);
+            case 2:
+                return player.getProperty(PlayerProperty.PROP_PLAYER_GEOCULUS);
+            case 3:
+                return player.getProperty(PlayerProperty.PROP_PLAYER_ELECTROCULUS);
+            case 4:
+                return player.getProperty(PlayerProperty.PROP_PLAYER_DENDROCULUS);
+            default:
+                return 0;
+        }
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/managers/SotSManager.java]: can be replaced with a modern return switch statement

File: src/main/java/emu/grasscutter/data/excels/StatuePromoteData.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        return costItems;
+    }
+
+    public int[] getRewardIdList() {
+        return rewardIdList;
+    }
+
+    public int getStamina() {
+        return stamina;
+    }
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/data/excels/StatuePromoteData.java]: all of these can be replaced with `@Getter` (lombok annotation)

File: src/main/java/emu/grasscutter/game/managers/SotSManager.java [hunk 3]
```diff
@@ -208,4 +214,140 @@ public void run() {
...
+            case 2:
+                player.setProperty(PlayerProperty.PROP_PLAYER_GEOCULUS, value);
+                break;
+            case 3:
+                player.setProperty(PlayerProperty.PROP_PLAYER_ELECTROCULUS, value);
+                break;
+            case 4:
+                player.setProperty(PlayerProperty.PROP_PLAYER_DENDROCULUS, value);
+                break;
+        }
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/managers/SotSManager.java]: can be replaced with a modern return switch statement

File: src/main/java/emu/grasscutter/game/managers/SotSManager.java [hunk 4]
```diff
@@ -208,4 +214,140 @@ public void run() {
...
+                return player.getProperty(PlayerProperty.PROP_PLAYER_ANEMOCULUS_LEVEL);
+            case 2:
+                return player.getProperty(PlayerProperty.PROP_PLAYER_GEOCULUS_LEVEL);
+            case 3:
+                return player.getProperty(PlayerProperty.PROP_PLAYER_ELECTROCULUS_LEVEL);
+            case 4:
+                return player.getProperty(PlayerProperty.PROP_PLAYER_DENDROCULUS_LEVEL);
+            default:
+                return 0;
+        }
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/managers/SotSManager.java]: can be replaced with a modern return switch statement

File: src/main/java/emu/grasscutter/game/managers/SotSManager.java [hunk 5]
```diff
@@ -208,4 +214,140 @@ public void run() {
...
+                break;
+            case 3:
+                player.setProperty(
+                        PlayerProperty.PROP_PLAYER_ELECTROCULUS_LEVEL, this.getCurrentLevel(cityId) + 1);
+                break;
+            case 4:
+                player.setProperty(
+                        PlayerProperty.PROP_PLAYER_DENDROCULUS_LEVEL, this.getCurrentLevel(cityId) + 1);
+                break;
+        }
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/managers/SotSManager.java]: can be replaced with a modern return switch statement

File: src/main/java/emu/grasscutter/game/managers/SotSManager.java [hunk 6]
```diff
@@ -208,4 +214,140 @@ public void run() {
...
+
+            // update max stamina and notify client
+            player.setProperty(
+                    PlayerProperty.PROP_MAX_STAMINA,
+                    player.getProperty(PlayerProperty.PROP_MAX_STAMINA) + nextStatuePromoteData.getStamina(),
+                    true);
+
+            // Add items to inventory
+            if (nextStatuePromoteData.getRewardIdList() != null) {
+                for (int rewardId : nextStatuePromoteData.getRewardIdList()) {
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/managers/SotSManager.java]:
```suggestion
                for (var rewardId : nextStatuePromoteData.getRewardIdList()) {
```

File: src/main/java/emu/grasscutter/game/props/PlayerProperty.java [hunk 7]
```diff
@@ -60,7 +60,15 @@ public enum PlayerProperty {
...
-    PROP_PLAYER_ONLINE_TIME(10047); // New; unknown/un-used.
+    PROP_PLAYER_ONLINE_TIME(10047), // New; unknown/un-used.
+    PROP_PLAYER_ANEMOCULUS(10048, 0), // Amount of Anemoculus collected [0, 65]
+    PROP_PLAYER_ANEMOCULUS_LEVEL(10049, 1, 10), // level of Anemoculus collected [1, 10]
+    PROP_PLAYER_GEOCULUS(10050, 0), // Amount of Geoculus collected [0, 131]
+    PROP_PLAYER_GEOCULUS_LEVEL(10051, 1, 10), // level of Geoculus collected [1, 10]
+    PROP_PLAYER_ELECTROCULUS(10052, 0), // Amount of Electroculus collected [0, 180]
+    PROP_PLAYER_ELECTROCULUS_LEVEL(10053, 1, 10), // level of Electroculus collected [1, 10]
+    PROP_PLAYER_DENDROCULUS(10054, 0), // Amount of Dendroculus collected [0, 270]
+    PROP_PLAYER_DENDROCULUS_LEVEL(10055, 1, 10); // level of Dendroculus collected [1, 10]
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/props/PlayerProperty.java]: are these actual player properties? if so, they should be considered in the virtual currency section of adding items to the player

**Review by KingRainbow44 — CHANGES_REQUESTED** [review]

File: src/main/java/emu/grasscutter/game/props/PlayerProperty.java (hunk 7 above)
Comment (KingRainbow44) [reply src/main/java/emu/grasscutter/game/props/PlayerProperty.java]: bigger issue, this aren't props found in-game, these should be relocated elsewhere

Comment (Phong940253) [reply src/main/java/emu/grasscutter/game/props/PlayerProperty.java]: Ok so I will add these props to another field, something like cityInfo

Comment (Hartie95) [reply src/main/java/emu/grasscutter/game/props/PlayerProperty.java]: PlayerProgress might be a good place to put them.
There you can probably put them into a map of the type CityInfo, for example like this:
```java
@Data
public class CityInfo {
    int cityId;
    int level = 0;
    int currentProgress = 0;
    
    public CityInto toProto(){
        return CityInfo.newBuilder()
             .setCityId(cityId)
             .setLevel(level)
             .setCrystalNum(currentProgress)
             .build()
    }
}
```

File: src/main/java/emu/grasscutter/data/excels/StatuePromoteData.java (hunk 2 above)
Comment (Hartie95) [reply src/main/java/emu/grasscutter/data/excels/StatuePromoteData.java]: `@Data` on the class level would probably be even better

**Review by KingRainbow44 — APPROVED** [review]

**Pull request merged**